File: leetcode/number-of-islands/python/number-of-islands.py
class Solution:
    def numIslands(self, grid):
        def dfsMark(graph, r, c):
            if min(r, c) < 0 or r == len(graph) or c == len(graph[0]) or graph[r][c] == "0":
                return  # base case
            if graph[r][c] == "1":
                graph[r][c] = "0"

            # 상하 좌우 탐색
            dfsMark(graph, r+1, c)
            dfsMark(graph, r-1, c)
            dfsMark(graph, r, c+1)
            dfsMark(graph, r, c-1)

        count = 0
        ROWS, COLS = len(grid), len(grid[0])
        for m in range(ROWS):
            for n in range(COLS):
                if grid[m][n] == "1":
                    dfsMark(grid, m, n)
                    count += 1
        return count

File: leetcode/number-of-islands/c/number-of-islands.c
void dfsMark(char** grid, int r, int c, int m, int* n);

int numIslands(char** grid, int gridSize, int* gridColSize) {
    int count = 0;

    for (int i = 0; i < gridSize; i++) {
        for (int j = 0; j < *gridColSize; j++) {
            if (grid[i][j] == '1') {
                dfsMark(grid, i, j, gridSize, gridColSize);
                count++;
            }
        }
    }
    return count;
}

void dfsMark(char** grid, int r, int c, int m, int* n) {
    if (
        r < 0 ||        
        c < 0 ||
        r == m ||
        c == *n ||
        grid[r][c] == '0'
    ) {
        return;
    }
    if (grid[r][c] == '1') {
        grid[r][c] = '0';
    }
    
    dfsMark(grid, r + 1, c, m, n);
    dfsMark(grid, r - 1, c, m, n);
    dfsMark(grid, r, c + 1, m, n);
    dfsMark(grid, r, c - 1, m, n);
}

